tensorwaves/interfaces.py

```diff
@@ -0,0 +1,102 @@
+"""Defines top level interfaces of tensorwaves.
+
+"""
+
+from abc import ABC, abstractmethod
+
+
+class Function(ABC):
+    """
+    Defines the interface of a callable function.
+    The parameters of the model are separated from the domain variables. This
+    follows the mathematical definition, in which a function defines its domain
+    and parameters. However specific points in the domain are not relevant.
+    Hence while the domain variables are the argument of the evaluation
+    (see :func:`~Function.__call__`), the parameters are controlled via a
+    getter and setter (see :func:`~Function.parameters`). The reason for this
+    separation is to facilitate the events when parameters have changed.
+    """
+    @abstractmethod
+    def __call__(self, dataset: dict) -> list:
+        """Evaluate the function.
+
+        Args:
+            dataset: a dictionary with domain variable names as keys and a
+            an list or `numpy.array` of  with a
+
+        Return:
+            list or :class:`numpy.array` of values
+        """
+
+    @property
+    @abstractmethod
+    def parameters(self) -> dict:
+        """
+        """
+
+    @parameters.setter
+    @abstractmethod
+    def parameters(self, new_parameters: dict):
+        """
+        """
+
+
+class Estimator(ABC):
```

Comment:
What is an 'estimator' precisely? I can't formulate it in another way than something like 'estimate of the discrepancy between model and data set'. That suggests there should be some link to both a model and a data set.

Reply:
Yeah its exactly that. But in an interface should not define any state, that why it is this way.
If data and model would be added to `__call__`  then we static property of the comparison is lost. So I don't know any better way

Reply:
What is the static property of the comparison?

Reply:
That the model and the data stay fixed for one instance of an estimator. Only parameters of the model (which are also the parameters of the estimator) are the things that will/can change